diff/diff.go

```diff
@@ -71,6 +77,15 @@ type Applier interface {
 	Apply(ctx context.Context, desc ocispec.Descriptor, mount []mount.Mount, opts ...ApplyOpt) (ocispec.Descriptor, error)
 }
 
+// WithCompressor sets the function to be used to compress the diff
+// stream.
+func WithCompressor(f func(dest io.Writer, mediaType string) (io.WriteCloser, error)) Opt {
+	return func(c *Config) error {
+		c.Compressor = f
```

Comment:
Without mediatype?

Reply:
@fuweid WDYT about the following API?

```golang
// WithCompressor sets the function to be used to compress the diff
// stream and the MediaType to be used for the compressed content.
func WithCompressor(mediaType string, f func(dest io.Writer) (io.WriteCloser, error)) Opt {
       return func(c *Config) error {
               c.MediaType = mediaType
               c.Compressor = f
               return nil
       }
}
```

Reply:
The previous one looked good to me. With `Compressor func(io.Writer, string) (io.WriteCloser, error)`

Reply:
@tonistiigi LGTY the following? This forces MediaType to be set explicitly as well as the actual MediaType used in the differ is passed to the compresor when compression happens. 

```golang
// WithCompressor sets the function to be used to compress the diff
// stream and the MediaType to be used for the compressed content.
func WithCompressor(mediaType string, f func(mediaType string, dest io.Writer) (io.WriteCloser, error)) Opt {
       return func(c *Config) error {
               c.MediaType = mediaType
               c.Compressor = f
               return nil
       }
}
```

Reply:
Well, `WithMediaType` already exists for this so duplication would be weird. I thought the point of adding it also to the compressor was that the compressor would do extra validation and return error if it doesn't understand the expected mediatype. Not very useful as you still pass them together with every call, but maybe would allow you to have one compressor instance that handles all mediatypes and you pass the same object with different calls with different mediatypes.

Reply:
I agree with that it duplicates but I think we need to make sure that the client explicitly specify the mediatype of the blob created by the compressor (as @fuweid points here).
So how about the following interface + adding a check in `Compare()` function to return an error when mediatype is unset but the compressor is set?

```golang
// WithCompressor sets the function to be used to compress the diff
// stream.
func WithCompressor(f func(dest io.Writer, mediaType string) (io.WriteCloser, error)) Opt {
	return func(c *Config) error {
		c.Compressor = f
		return nil
	}
}
```

Reply:
@tonistiigi @ktock Is it possible to allow customized compressor return meditype like this?

```go
Compressor func(dest io.Writer) (_ io.WriteCloser, mediaType string, _ error)
```

I was thinking that the combination of `(dest io.Writer, mediaType string)` in signature is to remind the caller that use the right mediatype here. I didn't realize about the `WithMediaType` option which can override the mediaType. Therefore, I am thinking about that compressor takes responsibility to return mediatype.

Reply:
@fuweid I don't think this is possible. The mediatype is already passed to the `Writer()` before compressor gets called.

Reply:
@fuweid @tonistiigi 

I think `WithCompressor(compressor func(dest io.Writer, mediaType string) (io.WriteCloser, error))` +  "adding a check to the differ to return an error when `config.MediaType` unset" will force the caller to use Compare() in the following style:

```golang
descr, err = differ.Compare(ctx, lower, upper,
	diff.WithMediaType(mediaType),
	diff.WithCompressor(compressor),
)
```

Thus I think we can remind the caller that it uses the right mediatype. The mediaType stored to the result descriptor is passed to the compressor so the compressor can confirm that it doesn't create invalid data.

We should add the following comment to the definition of `Compresor` option.

```golang
	// Compressor is a function to compress the diff stream
	// instead of the default gzip compressor. Differ passes
	// the MediaType of the target diff content to the compressor.
	// When using this config, MediaType must be specified as well.
	Compressor func(dest io.Writer, mediaType string) (io.WriteCloser, error)
```

Reply:
@fuweid I tried the following interface but it seems impossible because the writer of the content store (`store.Writer()`) requires to [determine the mediatype of the contents before starting to write](https://github.com/containerd/containerd/blob/1292c50fe99e9deb01155edef2ec9062305f311c/diff/walking/differ.go#L90-L97).

```golang
Compressor func(dest io.Writer) (_ io.WriteCloser, mediaType string, _ error)
```

Instead, I would like to make it the caller's responsibility to specify the target mediatype (config.MediaType) of the final content created by the Compressor. So I updated the patch to enforce it by fixing Differ to return an error if no media type of the target content is explicitly specified but a compressor is used. PTAL?

Reply:
Yeah. Thanks! LGTM